### PR TITLE
Add LiteX equivalent of Xilinx FIFO_SYNC_MACRO to I2S

### PR DESCRIPTION
**Pull request by robertszczepanski**
This PR adds LiteX equivalent of Xilinx `FIFO_SYNC_MACRO` which is used in few SoC cores. It is also replaced in I2S core if there is passed toolchain different than `vivado`. This provides support for I2S to toolchains different than Vivado. It has been tested with F4PGA and Vivado on Arty A7 and it works fine on both configurations.

### CODE REVIEW

**Comment by enjoy-digital** [comment]
Thanks @robertszczepanski. This looks good. Could you just move the core to `cores/ram` and call it `xilinx_fifo_sync_macro.py`?

File: litex/soc/cores/i2s.py [hunk 1]
```diff
@@ -250,50 +257,56 @@ def __init__(self, pads, fifo_depth=256, controller=False, master=False, concate
...
+                    i_RDEN        = fifo.rden,
+                    o_DO          = fifo.rd_d,
+                    o_RDCOUNT     = fifo.rdcount,
+                    o_RDERR       = fifo.rderr,
+                    o_WRCOUNT     = fifo.wrcount,
+                    o_WRERR       = fifo.wrerr,
+                )
+            else:
+                self.submodules.rx_fifo = fifo = FIFOSyncMacro("18Kb", data_width=fifo_data_width,
+                                        almost_empty_offset=8, almost_full_offset=(512 - fifo_depth))
```
[the remark this answers is not in the document]
Comment (robertszczepanski) [reply litex/soc/cores/i2s.py:299]: Sure, I'll work on it.

**Comment by robertszczepanski** [comment]
Hi @enjoy-digital, I applied some changes, please take a look and let me know if there is still something to improve.

**Comment by enjoy-digital** [comment]
Thanks @robertszczepanski, this is fine for me now:
- I2S code is simplified (by avoiding direct FIFO instance/duplication).
- We now have an equivalent of Xilinx FIFO_SYNC_MACRO in LiteX, which can be useful for simulation or use the code on other FPGAs.

So these are already nice improvements on the I2S core regarding simulation/portability and also easy reuse of Xilinx FIFO_SYNC_MACRO for other cases.

@bunnie: Could you also have a look and provide feedback since you are the author of the code? If this is also fine for you, we could merge.

**Review by bunnie — COMMENTED** [review]
Sorry for the delay - I only became aware of the review request until yesterday.

Generally looks good to me but the changes are extensive enough that I'd feel better if I compiled this to hardware and checked it. I'll try to do that in the next 1-2 days -- right now I'm traveling and the good news is I did bring the kit to check things like this but the bad news is I'm at an airport and they security will probably give me a weird look if I unpack the hardware tools in the plane.

**Comment by bunnie** [comment]
Just tested it on real hardware -- seems to work great. Thanks!

**Comment by bunnie** [comment]
Just tested it on real hardware -- seems to work great. Thanks!

**Comment by enjoy-digital** [comment]
Great, thanks @bunnie. (It seems you've been discrete enough to avoid issues with the security at the airport :))

Thanks @robertszczepanski for the work and modifications, we can now merge this.

**Pull request merged**